src/main/java/com/nephest/battlenet/sc2/discord/GuildEmojiStore.java

```diff
@@ -0,0 +1,63 @@
+package com.nephest.battlenet.sc2.discord;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Component;
+import com.nephest.battlenet.sc2.model.BaseLeague;
+import com.nephest.battlenet.sc2.model.Race;
+import discord4j.core.event.domain.guild.EmojisUpdateEvent;
+import discord4j.core.event.domain.interaction.InteractionCreateEvent;
+import discord4j.core.object.entity.GuildEmoji;
+import discord4j.core.retriever.EntityRetrievalStrategy;
+import reactor.core.publisher.Mono;
+
+@Component
+public class GuildEmojiStore {
+
+    public Optional<String> getGuildLeagueEmoji(Map<String, String> guildEmojis, BaseLeague.LeagueType league) {
+        return Optional.ofNullable(guildEmojis.get(league.getName().toLowerCase(Locale.ENGLISH)));
+    }
+
+    public Optional<String> getGuildRaceEmoji(Map<String, String> guildEmojis, Race race) {
+        return Optional.ofNullable(guildEmojis.get(race.getName().toLowerCase(Locale.ENGLISH)));
+    }
+
+    @Cacheable(cacheNames = "discord-guild-emoji", key="#evt.getInteraction().getGuildId().get()?.asLong()")
+    public Map<String, String> getGuildEmojis(InteractionCreateEvent evt) {
+        return evt.getInteraction()
```

Comment:
I feel like this reactor/stream chain can be improved by splitting the tasks. 
There are 2 tasks here
* Collection of target emoji names
* Collection of emojis themselves

IMO the reactor chain can be more efficient and straightforward if the first task(target emoji names) is moved to an unmodifiable static `Set<String>`. This way we remove unnecessary map operations(replace and remove), and we are free to use a simple operator chain for emojis like `.flatMapMany().filter(<using static set from the previous step>).collectMap()`.

Reply:
Ah, very good point! This will help with both efficiency and readability.